### ISSUE TITLE
signal overriding reset state problem

### ISSUE THREAD

**Issue by krcb197** [SystemRDL/PeakRDL-html#26]
### Problem
The following system RDL code shows the case of a field whose reset value is based on a signal. The use case for this could be a special pin on the device, normally pulled high internally but when grounded externally causes the device to enter a special mode after reset. I believe this code is legal as there is an example in the [SystemRDL 2.0 Specification](https://www.accellera.org/images/downloads/standards/systemrdl/SystemRDL_2.0_Jan2018.pdf) which is very similar, see 8.3.2.

```
signal { signalwidth=1; } reset_state_override;

addrmap external_reset_change {

    reg {
        field {
            sw=rw;
            hw=r;
        } bar = 0x0;
    } foo;

    foo.bar -> reset = reset_state_override;
};
```
If you attempt to build the HTML from this code you will get the following final error: 
```TypeError: %x format: an integer is required, not SignalNode```

I believe this comes from the following line in the reg_base.html template:
```{{"0x%x" % field.get_property("reset")}}```
where the reset property of the field is converted to an integer, where as the reset property is a signal in this case.

### Possible Solution

I have hacked my way around the issue to get the code running by doing the following, however, there is a still a bit not quite right.

#### Template change
I made the following change in  reg_base.html
```
{%- if field.get_property("reset") != None -%}
    {%- if isinstance(field.get_property("reset"),SignalNode) -%}
      {{ field.get_property("reset").inst_name }}
    {%- else -%}
      {{"0x%x" % field.get_property("reset")}}
    {%- endif -%}
{%- else -%}
-
{%- endif %}
```
#### Change to the RAL dictionary

Once I had "fixed" the template I found a follow-up problem with the JSON export of the RAL data which I fixed by changing this code in the exporter.py ```visit_addressable_node``` method

```python

if "reset" in field.list_properties():
    field_reset = field.get_property("reset")
    if isinstance(field_reset, SignalNode):
        ral_feeld_reset = field_reset.inst_name
    else:
        ral_feeld_reset = BigInt(field_reset)
else:
    ral_feeld_reset = BigInt(0)

ral_field = {
    'name' : field.inst.inst_name,
    'lsb'  : field.inst.lsb,
    'msb'  : field.inst.msb,
    'reset': ral_feeld_reset,
    'disp' : 'H'
}

```

#### The "final" problem

With my hack whilst the code now builds there it still is not quite right as the Decode of the field in the resultant webpage is node quite showing right, see below:
![image](https://user-images.githubusercontent.com/34693973/128559214-06120378-e77f-4590-a657-0e0aff8bee07.png)

### Next Steps

I would be quite happy to have a go at finalising the bug fix by submitting a pull request. Though I need a bit of guidance about how to solve the "final" problem.

**Comment by amykyta3** [comment]
Thanks for pointing this one out. As you've noticed, the HTML exporter doesn't support component references too well yet. I was actually wondering how long it would take for someone to notice :smile:.

I think your solution is headed in the right direction. Since the actual value of the reference is not knowable, I think the best compromise is to just have the "decoded" value box get initialized to 0. It would probably be good to also emit a sub-heading for the signal itself and its documentation. Also recall that the reset value can also reference a `field` component. In that case it would make sense for the HTML generator to create a link to it.

If you want, submit your changes as a PR and I can use it as a starting point. There are a few other places where references can be used so I'd like to take a closer look at this. (I may not be able to get to it until ~September since I'll be traveling)

**Comment by krcb197** [comment]
That is a good point, I had not realised that at `field` component can be the reset control to another field, I will take another look at the specification. That may require some additional test cases on my project.

I will give suggesting a fix, watch this space

**Comment by krcb197** [comment]
I have proposed a solution see the draft pull request, all feedback is welcome.

**Comment by amykyta3** [comment]
Merged your PR into a dev branch. I'll use what you started and fixup remaining support for references. Thanks!

**Comment by amykyta3** [comment]
Fixed in v2.1.2

**Comment by rakesh2215** [comment]
I am hitting this issue for PeakRDL-ipxact. 
Any pointers to import the fix from PeakRDL-html will be helpful.

  File "exporter.py", line 463, in add_field
    self.add_value(reset_el, self.ns + "value", self.hex_str(reset))
  File "exporter.py", line 251, in hex_str
    return "'h%x" % v 
TypeError: %x format: an integer is required, not SignalNode
